src/lib/CofEnv.ml

```diff
@@ -4,55 +4,108 @@ module D = Domain
+  }
+
+type consistent_env =
+  { classes : D.dim UF.t;
+    (** equivalence classes of dimensions from reduced cofibrations *)
 
-    status : [`Consistent | `Inconsistent];
-    (** a cache which must be maintained *)
+    true_vars : VarSet.t;
 
-    true_vars : VarSet.t
+    unreduced_joins : D.cof list list;
+    (** a stack of unreduced joins, each represented by a list of cofibrations *)
   }
 
-let init () =
-  {classes = UF.init ~size:100;
-   true_vars = VarSet.empty;
-   cof = Cof.top;
-   status = `Consistent}
+type env = [ `Consistent of consistent_env | `Inconsistent ]
+
+let init () = `Consistent
+    {classes = UF.init ~size:100;
+     true_vars = VarSet.empty;
+     unreduced_joins = []}
 
-let inconsistent =
-  {classes = UF.init ~size:0;
-   cof = Cof.bot;
-   true_vars = VarSet.empty;
-   status = `Inconsistent}
+let inconsistent = `Inconsistent
 
-let status env = env.status
+let already_inconsistent =
+  function
+  | `Consistent _ -> `Consistent
+  | `Inconsistent -> `Inconsistent
 
 let find_class classes r =
   try UF.find r classes with _ -> r
 
-let unreduced_assumptions env =
-  env.cof
+(* minimum requirement to do the search *)
+module type SEQ =
+sig
+  (* The type of the result of the search. *)
+  type t
+
+  (* The default value for vacuous cases. Should be the same as `seq id []`. *)
+  val vacuous : t
 
+  (* The sequencing operator. Technically, we can demand `seq' : t list -> t` instead
+   * and the current `seq f l` would be `seq' (map f l)`. However, `List.for_all` and
+   * `CoolBasis.Monad.Util.app` directly fit into this type. *)
+  val seq : ('a -> t) -> 'a list -> t
+end
 
-module Inversion :
+module rec Search : functor (M : SEQ) ->
 sig
-  type local =
-    {classes : D.dim UF.t;
-     true_vars : VarSet.t}
+  (* Search all branches assuming more cofibrations *)
+  val left_invert : env -> D.cof list -> (reduced_env -> M.t) -> M.t
 
-  val left : local -> D.cof list -> D.cof -> bool
+  (* Search all branches assuming more cofibrations *)
+  val left_invert' : consistent_env -> D.cof list -> (reduced_env -> M.t) -> M.t
+end
+  =
+  functor (M : SEQ) ->
+  struct
+    let left_invert' env phis cont =
+      let rec go ({classes; true_vars; unreduced_joins} as env) =
+        function
+        | [] ->
+          begin
+            match unreduced_joins with
+            | [] -> cont {classes; true_vars}
+            | (psis :: unreduced_joins) ->
+              M.seq (fun psi -> go {env with unreduced_joins} [psi]) psis
+          end
+        | (phi :: phis) ->
+          match phi with
+          | Cof.Var v ->
+            go {env with true_vars = VarSet.add v true_vars} phis
+          | Cof.Cof phi ->
+            match phi with
+            | Cof.Meet psis ->
+              go env (psis @ phis)
+            | Cof.Join psis ->
+              let env = {env with unreduced_joins = psis :: unreduced_joins} in
+              if Test.simple env Cof.bot then M.vacuous else go env phis
```

Comment:
@jonsterling This is where `Test` is used. Now, thinking about it, perhaps we should just **continue the search?!**

Reply:
if we continue the search, won’t we waste a lot of effort checking definitional equivalence of large terms?

Reply:
I guess we need deep learning to decide whether we should run deep learning to decide whether we should check consistency...